### PR TITLE
s4742286 ISIC Segmentation with IUNet

### PR DESCRIPTION
**Pull request by dkasumagic**
Hello Shakes,

The ISIC2018 dataset is an image directory consisting of a large sample of RGB 318x512 images of lesions on the skin, its segmentation masks of the same size, and another folder of the same kind as the first, but no segmentation masks. The Improved UNet, or IUNet, is a CNN that passes through RGB images, convolves it and downsamples, then upsamples and returns its segmentation mask that should hopefully contain the legion inside while containing as little healthy skin as possible.

Files included in this request are:
`dataset.py`: Contains the loading functions of the CustomISICDataset
`modules.py`: Contains the model itself, along with the dice_loss function
`train.py`: Trains and evaluates the model, then saves it. Produces relevant figures and images for loss as well.
`predict.py`: Loads the model from `train.py`, takes random numbers and proceeds to plot the model against the ground truth. 
                   Also calculates the dice score.

Thank you for your time,
Daniel Kasumagic
s4742286

### CODE REVIEW

**Review by dkasumagic — COMMENTED** [review]
Everything looks like the most updated version.

**Comment by dkasumagic** [comment]
Boy was that a struggle to get to the end there. I was told by a tutor that 3pm was the submission time, and then I got the notification that it was due at 11:59pm today, so I couldn't just give up. The last commit was fixing the names of the GroupedResultsComparison files. Turns out I included a space in their names that was screwing it up.

I did end up managing to get a dice score above 0.8. Thank you and have a nice day.

**Comment by wangzhaomxy** [comment]
This is an initial inspection, no action is required at this point.

- Wrong files path. All student constructed document should be packed in one document under the ‘recognition’ document.
- Algorithm solves the problem appropriately. Average Dice Coefficient in validating dataset: 0.99, might be for the data leakage. Predicted mask output is not binary one, so looks a little bit blur. Other parts are ok.
- Using Pytorch to construct the Improved U-Net models and functions. However, the model body is roughly an improved U-Net model(filters start with 8 and end up with 128) instead of the improved U-Net in the paper(filters start with 16 and end up with 256). Just train/validation datasets, rather than the train/validation/test datasets. Other parts are ok.
- Good coding design. Good handwriting code.
- Sufficient docstring and comment in the code.
- Easy difficulty problem, no higher difficulty marks.
- Wrong pull request branch, from local:student_document to shakes76:main

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present in wrong directory, needs to be in own folder -2
Shows no Usage -1 & Demo & Visualisation & Data usage 
Module present
Commenting
No Data leakage
Difficulty: Easy -10
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in InCorrect Branch) -2
Feedback required, move code to own folder and restore repo README -2
Request Description good

**Comment by shakes76** [comment]
Correct the folder as requested above to recieve additional +2 feedback marks to allow merge.

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks granted.

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks granted.

**Comment by dkasumagic** [comment]
Wait hold on, I only had 4 days to fill this out? I’ve been busy with work, I got 0 notifications about github, and now I’ve just lost multiple marks?

**Comment by wangzhaomxy** [comment]
> Wait hold on, I only had 4 days to fill this out? I’ve been busy with work, I got 0 notifications about github, and now I’ve just lost multiple marks?

"Feedback required, move code to own folder and restore repo README -2"
Just fix these two things following Shakes' instructions. I'll give you back 2 marks if you can finish it by the end of today.

Cheers.
